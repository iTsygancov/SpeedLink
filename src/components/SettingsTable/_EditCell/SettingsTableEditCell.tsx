import { Button } from "@/components/ui/button";
import {
  DropdownMenu,
  DropdownMenuContent,
  DropdownMenuItem,
  DropdownMenuTrigger
} from "@/components/ui/dropdown-menu";
import { TableCell } from "@/components/ui/table";
import { Shortcut } from "@/types";
import { Check, MoreHorizontal, X } from "lucide-react";

type SettingsTableEditCellProps = {
  handleCloseEditShortcuts: () => void;
  handleEditShortcuts: (item: Shortcut, itemIndex: number) => void;
  handleSaveShortcut: (item: Shortcut, itemIndex: number) => void;
  isInEditMode: boolean;
  item: Shortcut;
  itemIndex: number;
  setCurrentCommand: (item: Shortcut) => void;
  setIsDialogOpen: (isDialogOpen: boolean) => void;
};

const SettingsTableEditCell = ({
  handleCloseEditShortcuts,
  handleEditShortcuts,
  handleSaveShortcut,
  item,
  itemIndex,
  isInEditMode,
  setCurrentCommand,
  setIsDialogOpen
}: SettingsTableEditCellProps) => {
  return (
    <TableCell className='p-0 pr-4 text-right'>
      {item.canEdit ? (
        <>
          <Button
            className='mr-4 h-7 px-4'
            onClick={() => handleSaveShortcut(item, itemIndex)}
            size='sm'
          >
            <Check size={16} />
          </Button>
          <Button
            variant='outline'
            size='sm'
            className='h-7 px-4'
            onClick={handleCloseEditShortcuts}
          >
            <X size={16} />
          </Button>
        </>
      ) : (
        <>
          {!isInEditMode && (
            <DropdownMenu>
              <DropdownMenuTrigger className='focus-within:outline-offset-2 focus-within:outline-primary'>
                <MoreHorizontal size={24} className='cursor-pointer ' />
              </DropdownMenuTrigger>
              <DropdownMenuContent>
                <DropdownMenuItem
                  onClick={() => handleEditShortcuts(item, itemIndex)}
                >
                  Edit
                </DropdownMenuItem>
                <DropdownMenuItem
                  className='text-red-500'
                  onClick={() => {
                    setCurrentCommand(item);
                    setIsDialogOpen(true);
                  }}
                >
                  Delete
                </DropdownMenuItem>
              </DropdownMenuContent>
            </DropdownMenu>
          )}
        </>
      )}
    </TableCell>
  );
};

export default SettingsTableEditCell;
