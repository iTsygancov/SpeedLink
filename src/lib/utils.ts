import { type ClassValue, clsx } from "clsx";
import { twMerge } from "tailwind-merge";

export function cn(...inputs: ClassValue[]) {
  return twMerge(clsx(inputs));
}

export function generateAlphanumericKeysArray() {
  const keysArray = [];

  // Generate lowercase alphabetic keys
  for (let i = 65; i <= 90; i++) {
    keysArray.push(String.fromCharCode(i));
  }

  // Generate numeric keys
  for (let i = 48; i <= 57; i++) {
    keysArray.push(String.fromCharCode(i));
  }

  return keysArray;
}
