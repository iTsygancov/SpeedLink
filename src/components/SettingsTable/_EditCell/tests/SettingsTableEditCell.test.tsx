import SettingsTableEditCell, {
  SettingsTableEditCellProps
} from "../SettingsTableEditCell";
import { initialCommand } from "@/test/mockData";
import { expectToMatchSnapshot, renderTree } from "@/test/utils";
import { describe, it, vi } from "vitest";

const initialProps = {
  handleCloseEditShortcuts: vi.fn(),
  handleEditShortcuts: vi.fn(),
  handleSaveShortcut: vi.fn(),
  isInEditMode: false,
  item: initialCommand,
  itemIndex: 0,
  setCurrentCommand: vi.fn(),
  setIsDialogOpen: vi.fn()
};

const testComponentWithProps = (props: SettingsTableEditCellProps) => {
  expectToMatchSnapshot(renderTree(<SettingsTableEditCell {...props} />));
};

describe("SettingsTableEditCell", () => {
  it("should render edit buttons", () => {
    testComponentWithProps(initialProps);
  });

  it("should render dropdown", () => {
    testComponentWithProps({
      ...initialProps,
      item: {
        ...initialProps.item,
        canEdit: false
      }
    });
  });

  it("should render cell without edit buttons and dropdown", () => {
    testComponentWithProps({
      ...initialProps,
      item: {
        ...initialProps.item,
        canEdit: false
      },
      isInEditMode: true
    });
  });
});
