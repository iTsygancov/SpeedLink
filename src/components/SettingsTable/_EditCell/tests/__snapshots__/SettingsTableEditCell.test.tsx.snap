// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SettingsTableEditCell > should render cell without edit buttons and dropdown 1`] = `
<td
  className="align-middle [&:has([role=checkbox])]:pr-0 p-0 pr-4 text-right"
/>
`;

exports[`SettingsTableEditCell > should render dropdown 1`] = `
<td
  className="align-middle [&:has([role=checkbox])]:pr-0 p-0 pr-4 text-right"
>
  <button
    aria-expanded={false}
    aria-haspopup="menu"
    className="focus-within:outline-offset-2 focus-within:outline-primary"
    data-state="closed"
    disabled={false}
    id="radix-:r2:"
    onKeyDown={[Function]}
    onPointerDown={[Function]}
    type="button"
  >
    <svg
      className="cursor-pointer "
      fill="none"
      height={24}
      stroke="currentColor"
      strokeLinecap="round"
      strokeLinejoin="round"
      strokeWidth={2}
      viewBox="0 0 24 24"
      width={24}
      xmlns="http://www.w3.org/2000/svg"
    >
      <circle
        cx="12"
        cy="12"
        r="1"
      />
      <circle
        cx="19"
        cy="12"
        r="1"
      />
      <circle
        cx="5"
        cy="12"
        r="1"
      />
    </svg>
  </button>
</td>
`;

exports[`SettingsTableEditCell > should render edit buttons 1`] = `
<td
  className="align-middle [&:has([role=checkbox])]:pr-0 p-0 pr-4 text-right"
>
  <button
    aria-expanded={false}
    aria-haspopup="menu"
    className="focus-within:outline-offset-2 focus-within:outline-primary"
    data-state="closed"
    disabled={false}
    id="radix-:r0:"
    onKeyDown={[Function]}
    onPointerDown={[Function]}
    type="button"
  >
    <svg
      className="cursor-pointer "
      fill="none"
      height={24}
      stroke="currentColor"
      strokeLinecap="round"
      strokeLinejoin="round"
      strokeWidth={2}
      viewBox="0 0 24 24"
      width={24}
      xmlns="http://www.w3.org/2000/svg"
    >
      <circle
        cx="12"
        cy="12"
        r="1"
      />
      <circle
        cx="19"
        cy="12"
        r="1"
      />
      <circle
        cx="5"
        cy="12"
        r="1"
      />
    </svg>
  </button>
</td>
`;
